### ISSUE TITLE
Heapmem does not return correctly aligned addresses

### ISSUE THREAD

**Issue by akindestam** [contiki-ng/contiki-ng#999]
In heapmem.c the ```heap_base``` buffer is allocated as an array of ```char```:
```
static char heap_base[HEAPMEM_ARENA_SIZE];
```
 Since ```char``` has an alignment of only 1 byte the compiler is free to start this array at any byte (for reproductions sake: I am building on ARM with GCC and -Os). As heapmem seems to align addresses against the start of this buffer, you can see how this gives bad results when you have ```HEAPMEM_ALIGNMENT``` set to 4.

A solution would be to define using ```__attribute__((aligned(HEAPMEM_ALIGNMENT)))``` at least on compilers which support this GCC extension:
```
__attribute__((aligned(HEAPMEM_ALIGNMENT)))
static char heap_base[HEAPMEM_ARENA_SIZE];
```

<strike>Allocating the array out of a type with a suitable alignment might be another potentially more portable way, although making that configurable might be tricky. Example for ```HEAPMEM_ALIGNMENT == 4```:
```
static uint32_t heap_base_buf[1 + (HEAPMEM_ARENA_SIZE - 1)/sizeof(uint32_t)]; // round up
static char * heap_base = (char*)&heap_base_buf[0];
```
One solution for the configurability might simply be to allocate the buffer out of a type with a suitably large alignment and then allow HEAPMEM_ALIGNMENT to be configured up to that.</strike>
EDIT: There is ```CC_ALIGN``` for portability.

**Comment by atiselsts** [comment]
There's `CC_ALIGN` macro that could be used instead of explicitly invoking GCC attributes.

**Comment by akindestam** [comment]
> There's `CC_ALIGN` macro that could be used instead of explicitly invoking GCC attributes.

Ah, thanks for pointing that out!

**Comment by g-oikonomou** [comment]
Fixed in #1001
